hbase-client/src/main/java/org/apache/hadoop/hbase/client/RpcRetryingCallerImpl.java

```diff
@@ -140,21 +141,30 @@ public T callWithRetries(RetryingCallable<T> callable, int callTimeout)
         if (tries >= maxAttempts - 1) {
           throw new RetriesExhaustedException(tries, exceptions);
         }
-        // If the server is dead, we need to wait a little before retrying, to give
-        // a chance to the regions to be moved
-        // get right pause time, start by RETRY_BACKOFF[0] * pauseBase, where pauseBase might be
-        // special when encountering an exception indicating the server is overloaded.
-        // see #HBASE-17114 and HBASE-26807
-        long pauseBase =
-          HBaseServerException.isServerOverloaded(t) ? pauseForServerOverloaded : pause;
-        expectedSleep = callable.sleep(pauseBase, tries);
-
-        // If, after the planned sleep, there won't be enough time left, we stop now.
-        long duration = singleCallDuration(expectedSleep);
-        if (duration > callTimeout) {
-          String msg = "callTimeout=" + callTimeout + ", callDuration=" + duration + ": "
-            + t.getMessage() + " " + callable.getExceptionMessageAdditionalDetail();
-          throw (SocketTimeoutException) new SocketTimeoutException(msg).initCause(t);
+        if (t instanceof RpcThrottlingException) {
+          RpcThrottlingException rpcThrottlingException = (RpcThrottlingException) t;
+          expectedSleep = rpcThrottlingException.getWaitInterval();
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Sleeping for {}ms after catching RpcThrottlingException", expectedSleep,
+              rpcThrottlingException);
+          }
+        } else {
+          // If the server is dead, we need to wait a little before retrying, to give
+          // a chance to the regions to be moved
+          // get right pause time, start by RETRY_BACKOFF[0] * pauseBase, where pauseBase might be
+          // special when encountering an exception indicating the server is overloaded.
+          // see #HBASE-17114 and HBASE-26807
+          long pauseBase =
+            HBaseServerException.isServerOverloaded(t) ? pauseForServerOverloaded : pause;
+          expectedSleep = callable.sleep(pauseBase, tries);
+
+          // If, after the planned sleep, there won't be enough time left, we stop now.
+          long duration = singleCallDuration(expectedSleep);
+          if (duration > callTimeout) {
+            String msg = "callTimeout=" + callTimeout + ", callDuration=" + duration + ": "
+              + t.getMessage() + " " + callable.getExceptionMessageAdditionalDetail();
+            throw (SocketTimeoutException) new SocketTimeoutException(msg).initCause(t);
+          }
```

Comment:
So this method here is where we do retries. The fact that we're throwing a SocketTimeoutException means that the retries will stop.

I think we should move this logic outside the if. Let's say the wait interval is 10s, but operation timeout is 5s. Honoring the wait interval will cause us to exceed the timeout, which is not good. At the very least we should do a Math.min, but we've also been told by the server that the quota won't be available until 10s. So we know that if we retry in 5s (due to a Math.min), we will fail. So with that said, we should just move the logic outside the if so that we throw an exception if the wait interval extends beyond operation timeout.

Regarding the quota -- if the server says 10s, i dont think it will ever be ready in, say, 8s. The wait interval is calculated based on the the rate of quota refill, so it should be accurate as a lower threshold. What we can't be sure of is if the quota will _actually_ be ready in 10s (may be longer). If there were only 1 caller, the wait interval would be a great indicator of exactly when to retry. But in the case of N > 1 callers, those other callers may also be consuming the quota and the client may retry in 10s but it's still been saturated in the interim and wasn't able to refill enough to serve the request.